lib/public_key/src/pubkey_cert.erl

```diff
@@ -558,82 +637,829 @@ root_cert(Name, Opts) ->
+			      | Rest], ValidationState, ExistBasicCon,
+		    SelfSigned, UserState, VerifyFun) ->
+    NewValidationState = ValidationState#path_validation_state{policy_constraint_ext = Ext},
+    validate_extensions(OtpCert, Rest, NewValidationState, ExistBasicCon,
+			SelfSigned, UserState, VerifyFun);
+validate_extensions(OtpCert, [#'Extension'{extnID = ?'id-ce-policyMappings'} = Ext
+                             | Rest], ValidationState, ExistBasicCon,
+		    SelfSigned, UserState, VerifyFun) ->
+    NewValidationState = ValidationState#path_validation_state{policy_mapping_ext = Ext},
+    validate_extensions(OtpCert, Rest, NewValidationState, ExistBasicCon,
+			SelfSigned, UserState, VerifyFun);
+validate_extensions(OtpCert, [#'Extension'{extnID = ?'id-ce-inhibitAnyPolicy'} = Ext
+			      | Rest], ValidationState, ExistBasicCon,
+		    SelfSigned, UserState, VerifyFun) ->
+    NewValidationState = ValidationState#path_validation_state{policy_inhibit_ext = Ext},
+    validate_extensions(OtpCert, Rest, NewValidationState, ExistBasicCon,
+			SelfSigned, UserState, VerifyFun);
+
+validate_extensions(OtpCert, [#'Extension'{} = Extension | Rest],
+		    ValidationState, ExistBasicCon,
+		    SelfSigned, UserState0, VerifyFun) ->
+    UserState = verify_fun(OtpCert, {extension, Extension}, UserState0, VerifyFun),
+    validate_extensions(OtpCert, Rest, ValidationState, ExistBasicCon, SelfSigned,
+			UserState, VerifyFun).
+
+handle_last_cert(OtpCert, #path_validation_state{last_cert = true,
+                                                 user_initial_policy_set = PolicySet,
+                                                 valid_policy_tree = Tree} = ValidationState0) ->
+    TBSCert = OtpCert#'OTPCertificate'.tbsCertificate,
+    Extensions =
+        extensions_list(TBSCert#'OTPTBSCertificate'.extensions),
+    %% 6.5.1 b
+    ValidationState  =
+        case select_extension(?'id-ce-policyConstraints', Extensions) of
+            undefined ->
+                ValidationState0;
+            #'Extension'{extnValue = #'PolicyConstraints'{requireExplicitPolicy = 0}} ->
+                ValidationState0#path_validation_state{explicit_policy = 0};
+            _  ->
+                ValidationState0
+    end,
+    ValidTree = policy_tree_intersection(PolicySet, Tree),
+    validate_policy_tree(OtpCert,
+                         ValidationState#path_validation_state{valid_policy_tree = ValidTree});
+handle_last_cert(_, ValidationState) ->
+    ValidationState.
+
+
+%%====================================================================
+%% Policy handling
+%%====================================================================
+%% Start initialization  RFC 5280 Section 6.1.2 ----------------------
+
+%%  6.1.2 d, e, f: If <indicator> is set, then the initial value
+%%  is 0, otherwise the initial value is n+1. (N = max path length)
+policy_indicator(_, true) ->
+    0;
+policy_indicator(N, false) ->
+   N + 1.
+
+policy_set(Opts, Default) ->
```

Comment:
should we remove potential duplicates from user input?
RFC operates on `set` we are using a `list` ....

Reply:
A list is an easier input type. But we could change it to use the sets module internally to avoid the potential problem that the user provides duplicates.

Reply:
This is still an outstanding question!

Reply:
Is this not relevant anymore? please explain

Reply:
Duplicates in user input does not make any sense and even if duplicates would exist I  do not think they will create any incorrect results. Also we would probably need to convert from the set to list again internally to iterate over the set.